src/host/_stream.cpp

```diff
@@ -228,31 +291,91 @@ void WriteCharsLegacy(SCREEN_INFORMATION& screenInfo, const std::wstring_view& t
+                const auto ch = gsl::narrow_cast<char>(wch);
+                const auto result = MultiByteToWideChar(cp, MB_USEGLYPHCHARS, &ch, 1, &wch, 1);
+                if (result != 1)
+                {
+                    wch = 0;
+                }
+                if (wch)
+                {
+                    lastCharWrapped = _writeCharsLegacyUnprocessed(screenInfo, { &wch, 1 }, psScrollY);
+                }
                 break;
             }
+            }
 
-            // As a special favor to incompetent apps that attempt to display control chars,
-            // convert to corresponding OEM Glyph Chars
-            const auto cp = ServiceLocator::LocateGlobals().getConsoleInformation().OutputCP;
-            const auto ch = gsl::narrow_cast<char>(*it);
-            wchar_t wch = 0;
-            const auto result = MultiByteToWideChar(cp, MB_USEGLYPHCHARS, &ch, 1, &wch, 1);
-            if (result == 1)
+            if (io)
             {
-                _writeCharsLegacyUnprocessed(screenInfo, { &wch, 1 }, psScrollY);
+                if (wch)
+                {
+                    io->WriteUTF16({ &wch, 1 });
+                }
+                if (lastCharWrapped)
+                {
+                    io->WriteUTF8(" \r");
+                }
             }
+
+            ++it;
+        } while (it != end && controlCharPredicate(*it));
+    }
+}
+
+void WriteCharsVT(SCREEN_INFORMATION& screenInfo, const std::wstring_view& str)
+{
+    auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+
+    if (const auto io = gci.GetVtIo(&screenInfo))
+    {
+        using Mode = Microsoft::Console::VirtualTerminal::TerminalInput::Mode;
+
+        auto& terminalInput = gci.GetActiveInputBuffer()->GetTerminalInput();
+
+        // These two modes are ones that should always stay enabled when we're ConPTY.
+        // To figure out whether some VT sequence disabled them (primarily RIS = ESC c),
+        // we temporarily enable them, restore them, and check if they got reset in between.
+        const auto beforeFocusEvent = terminalInput.GetInputMode(Mode::FocusEvent);
+        const auto beforeWin32 = terminalInput.GetInputMode(Mode::Win32);
+        terminalInput.SetInputMode(Mode::FocusEvent, true);
+        terminalInput.SetInputMode(Mode::Win32, true);
+
+        screenInfo.GetStateMachine().ProcessString(str);
+
+        const auto afterFocusEvent = terminalInput.GetInputMode(Mode::FocusEvent);
+        const auto afterWin32 = terminalInput.GetInputMode(Mode::Win32);
+        terminalInput.SetInputMode(Mode::FocusEvent, beforeFocusEvent);
+        terminalInput.SetInputMode(Mode::Win32, beforeWin32);
+
+        const auto cork = io->Cork();
+        io->WriteUTF16(str);
+        if (!afterFocusEvent)
+        {
+            io->WriteUTF8("\033[?1004h");
+        }
+        if (!afterWin32)
+        {
+            io->WriteUTF8("\033[?9001h");
         }

```

Comment:
I think this is causing some problems with VT52 mode. Try executing this in a WSL bash shell:
```
printf "\ec\e[?2l\n\e<"
```
That's a hard reset, a switch to VT52 mode, a linefeed, and then a switch back to ANSI mode. It should just be the equivalent of a reset with a linefeed, but it ends up writing `?1004h?9001h` to the screen.

Reply:
Actually, it looks like it's not just VT52 mode. I'm also seeing an issue in ANSI mode that looks like it might be related.
```
printf "\ec\e[1\n2B"
```
That's a hard reset followed by `CUD` 12 (cursor down) with a linefeed embedded in the `CUD` sequence. That should move down 13 lines, but the linefeed ends up aborting the `CUD` sequence so it writes out `2B`. Maybe it's the linefeed that's the source of the problem?

Reply:
Yeah, I bet `printf` is line-buffered and so I inject the above sequences after each line feed.

To solve it properly, I need to record the positions of RIS and DECSET 1004 (focus event mode) to properly inject ConPTY's stuff in there. However, neither `AdaptDispatch`, nor even `OutputStateMachineEngine`, have any idea over the current offset in the to-be-parsed string and so this is quite painful to do. I wonder if passing parent pointers to the child classes is the only option... (= give `OutputStateMachineEngine` or `AdaptDispatch` a pointer to the `StateMachine`, so that it can tell it about the RIS/DECSET it encountered). I dislike that because it further taints the ownership "tree". Oh well...